📽️ [Step3] BoxOffice : Team Milo

sloDevz opened the pull request:
# Step3 PR

> 안녕하세요 제이크~! (@jryoun1) 열심히 하는 Team Milo 입니다. 어른이날 잘 보내셨나요? 저희에게는 4살 이하의 조카가 있는데 둘 다 휴일동안 조카 봐주느라 오히려 전 주보다 체력이 더 빠져버렸답니다~ 🤗💀 새삼 부모가 된다는건 얼마나 고행길일지 두려워 지는 휴일이었습니다 😌
저희 Team Milo의 이번 프로젝트 대주제는 ModernCollectionView 이기 때문에 이번 Step3에서는 ModernCollectionView를 집중적으로 패보는 시간이었습니다. 결과적으로 굉장히 만족스러운 수확이 있었던 Step이었어요!! 좋은 리뷰 부탁드립니다~~ @
> 

# ✅ 요구사항 & 체크리스트

---

- [x]  Step 2에서 구현한 네트워킹 기능을 통해 실제로 상품목록을 API 서버에 요청하여 불러옵니다.
- [x]  `어제`의 박스오피스를 볼 수 있는 화면을 구현합니다.
- [x]  리스트를 아래로 잡아끌어서 놓으면 리스트를 새로고침합니다 (당겨서 새로고침)
- [x]  처음 목록을 로드할 때, 사용자에게 빈 화면만 보여주는 대신, 로드 중임을 알 수 있게 해주세요.
- [x]  화면 상단에는 날짜를 표기합니다.
- [x]  리스트 형태로 박스오피스 정보를 표기합니다
- [x]  박스오피스 정보의 각 열에 표기할 필수정보
    - 맨 왼쪽에는 영화의 현재 등수를 표기합니다
        - 신규 영화면 등수 아래에 `신작`이라고 표기합니다
        - 기존 영화면 어제와 비교한 등락을 표기합니다
            - 순위 상승 : 빨간 화살표 + 등락 편차
            - 순위 하락 : 파란 화살표 + 등락 편차
            - 변동 없음 : `` 표기
- [x]  해당 일자의 관객수와 누적 관객수를 표기합니다
    - 숫자가 세 자리 이상 넘어가면 `,`를 활용하여 읽기 쉽도록 합니다. 예) 10,000

### Step3 핵심 경험

- [x]  Safe Area을 고려한 오토 레이아웃 구현
- [x]  Collection View의 활용
- [x]  Mordern Collection View 활용

---
# ✅ 구동화면

![Simulator Screen Recording - iPhone 14 Pro - 2023-05-09 at 17 32 35](https://user-images.githubusercontent.com/29590768/237040683-7cb8b798-8886-4d9f-9515-1687858583f7.gif)

---
# 🛠️ 구현 사항

### 🍭  Collection View 구현

`Compositional Layout`과 `DiffableDatasource`를 활용해서 Collection view를 구현했어요.

- `UICollectionLayoutListConfiguration` 을 활용해서 list 형식의 Collection view를 구현했어요.

```swift
let layout = UICollectionViewCompositionalLayout { section, layoutEnvironment in
            let configuration = UICollectionLayoutListConfiguration(appearance: .plain)

            return NSCollectionLayoutSection.list(using: configuration, layoutEnvironment: layoutEnvironment)
        }

return layout
```

- `UICollectionViewDiffableDatasource`를 활용했어요. 또한, `UICollectionViewCell`을 활용해서 cell에 악세서리[.disclosureIndicator()]를 넣어주었어요.

```swift
dataSource = UICollectionViewDiffableDataSource(collectionView: dailyBoxOfficeCollectionView )
        { collectionView, indexPath, movie in
            let cell = collectionView.dequeueReusableCell(withReuseIdentifier: DailyBoxOfficeCell.identifier,
                                                          for: indexPath) as? DailyBoxOfficeCell
            cell?.accessories = [.disclosureIndicator()]
            cell?.configure(with: movie)

            return cell
        }
```

### 🍭 API 호출시 로딩화면

- `UIActivityIndicatorView`를 활용하여 API 호출이 완료될 때까지의 대기 화면을 구현해요.

```swift
private var loadingIndicatorView = UIActivityIndicatorView(style: .large)

.
.
.

private func window() -> UIWindow {
        let scenes = UIApplication.shared.connectedScenes
        let windowScene = scenes.first as? UIWindowScene
        guard let window = windowScene?.windows.first else { return UIWindow() }

        return window
    }

private func showIndicatorview() {
        let window = window()
        loadingIndicatorView.frame = window.frame
        loadingIndicatorView.color = .systemBlue
        window.addSubview(loadingIndicatorView)
        loadingIndicatorView.startAnimating()
    }
```

- API의 요청이 완료된 후 성공적으로 데이터가 받아와졌을 때 로딩 화면을 숨겨요

```swift
private func hideIndicatorView() {
        let window = window()
        let indicatorView = window.subviews.first { $0 is UIActivityIndicatorView }
        guard let indicatorView else { return }
        indicatorView.removeFromSuperview()
    }
```

### 🍭 DateFormatter 활용

- `DateFormatter extension`, `String extension`을 활용했어요.

```swift
extension Date {

    static let yearMonthDaySplitDashFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"

        return formatter
    }()

    static func yesterDayDateConvertToString() -> String {
        let date = Date()
        guard let yesterDay = Calendar.current.date(byAdding: .day, value: -1, to: date) else { return "-" }
        let formatter = Date.yearMonthDaySplitDashFormatter

        return formatter.string(from: yesterDay)
    }

}
```

- 날짜들 중간에 들어가는 `“-”` 를 제거 하는 매서드를 `String extension`으로 따로 만들어서 썼어요.

```swift
extension String {

    func except(for characterSet: String) -> String {
        let charSet = CharacterSet(charactersIn: characterSet)
        let dashDeletedString = self.components(separatedBy: charSet).joined()

        return dashDeletedString
    }

}
```

---

# **⁉️ 궁금한 점**

## 🔨  옵셔널 바이딩이 실패할 경우의 리턴값

```swift
private func window() -> UIWindow {
        let scenes = UIApplication.shared.connectedScenes
        let windowScene = scenes.first as? UIWindowScene
        guard let window = windowScene?.windows.first else { return UIWindow() }

        return window
    }
```

- 해당 매서드는 windowScene에서 가장 첫번째 window를 받아오는 역할을 구현한 매서드에요.
- guard-let 으로 옵셔널 바인딩을 했고, 실패했을 경우 UIWindow 인스턴스를 만들어서 리턴해줘요.
- 실패할 경우, return한 UIWindow()는 원하는 Window 객체가 아니므로 아무 의미 없는 값이에요. 그러나, 런타임에서 에러가 나지 않도록 return 값을 넣어주었어요.
- ⁉️ 혹시 현업에서는 이러할 경우, 의미 없는 return 값을 주시나요, 아니면 error를 정의해서 error 핸들링을 해주시나요?

또 다른 예시는 아래와 같아요.

```swift
func convertStringToInt(number: String) -> Int {
        guard let number = Int(number) else { return 0 }
        return number
    }

func convertStringToInt(number: String) -> Int {
        guard let number = Int(number) else { return Int() }
        return number
    }
```

- guard-let 구문을 사용하는 여러 상황에서 옵셔널 바인딩이 실패했을 때 현업에서는 어떤 값들을 주로 리턴해주는지가 궁금합니다.
- 예를들어 guard-let String의 경우 “” 을 리턴하기도 하고 String() 이렇게 리턴하기도 했었는데 마땅한 기준을 정하기도 힘들고 이게 맞는지 의문이 들어서요 (⁉️ 상황마다 다를 것이라는 생각이 들긴하지만, 제이크는 이러할 경우, 어떤 식으로 처리를 하는 지 궁금합니다!)

## 🔨 dataSource.apply는 언제 호출하는게 바람직 한가?

- 저희는 DiffableDatasource를 활용해서 매번 현재의 snapshot을 생성해서 변경된 data를 UI에 업데이트해요!
- 아래의 코드에서 볼 수 있드시, 저희는 itemIdentifier를 담고 있는 container가 변경될 때마다 applySnapShot()을 해줘요.
- 그런데, 저희가 들었던 의문이 있었어요! ⁉️ 만약 컨테이너 속 하나의 DailyBoxOffice가 변경될 때마다 applySnapShot을 할 경우, 계속해서 여러 번의 applySnapShot이 호출되어 효율 측면에서 문제가 생기지 않을까 싶었어요.(reloadData()를 활용하지 않으므로, 더 효율적이기는 하지만..) 제이크가 보셨을 때, 이러한 UI update 방식이 어떨 지 궁금해요!

```swift
// 📚 DailyBoxOfficeViewController
private var movies = [DailyBoxOffice]() {
        didSet {
            applySnapShot()
        }
    }

.....
private func applySnapShot() {
        var snapShot = SnapShot()
        snapShot.appendSections([.main])
        snapShot.appendItems(movies)

        DispatchQueue.main.async {
            self.dataSource?.apply(snapShot)
        }
    }
```

### 🛠️ API로 부터 영화 디테일 정보를 요청하는 시점에 대하여

- 영화를 선택했을 때 그 영화에 대한 디테일 API를 요청할지? 아니면 처음 BoxOffice를 불러올때 영화 디테일 정보까지 한번에 호출해서 불러온 다음 저장을 해둬야 할지가 궁금합니다. 만약 상황에 따라 다르다면 제이크는 어떻게 생각하시는지 궁금해요!

sloDevz committed:
feat: BoxOffice Hashable 프로토콜 채택 및 이름 변경

longlivedrgn committed:
feat: DailyBoxOfficeCell property 정의 및 constant 정의하기

sloDevz committed:
feat: DailyBoxOfficeCell의 configureHierarchy, configureConstraints 메서드 정의

longlivedrgn committed:
feat: configure 함수 생성, audience count property 추가, 관객 수 생성하는 함수 추가

sloDevz committed:
feat: DailyBoxOfficeCell 의 Cell 컨텐츠 구성하는 메서드 작성

longlivedrgn committed:
feat: navigation controller 임베드 및 collection view 넣고, compositionalLayout 설정해주기

sloDevz committed:
feat: CollectionView AutoLayout 정의 DailyBoxOfficeCell 을 CollectionView에 등록

longlivedrgn committed:
feat: Datasource 구현하는 함수 생성 및 snapshot 생성 함수 추가

sloDevz committed:
feat: snapshot 호출 정의 및 DailyBoxOfficeCell 수정

longlivedrgn committed:
feat: Dateformatter를 활용해서 어제 날짜 가져오기

sloDevz committed:
feat: Network통신중 표시할 IndicatorView 구현

longlivedrgn committed:
feat: configureRefreshControl() 함수를 통해서 RefreshControl 구현하기

longlivedrgn committed:
feat: Constants 열거형 생성, extension 활용

sloDevz committed:
chore: String 매서드 개선 및 print코드 제거

longlivedrgn committed:
chore: 불필요한 메소드 및 개행 제거

sloDevz committed:
Merge branch 'step3' of https://github.com/longlivedrgn/ios-box-office into step3

longlivedrgn committed:
chore: storyboard 삭제하기

longlivedrgn committed:
chore: main file 제거하기

jryoun1 commented on BoxOffice/BoxOfficeViewController.swift:
```diff
@@ -1,17 +0,0 @@
-
-import UIKit
-
-final class BoxOfficeViewController: UIViewController {
```
불필요한 코드제거 좋네요 👍

jryoun1 commented on BoxOffice/Model/DTO/BoxOffice.swift:
```diff
@@ -30,7 +30,7 @@ struct BoxOfficeResult: Decodable {
 }
 
-struct DailyBoxOfficeList: Decodable {
+struct DailyBoxOffice: Decodable, Hashable {
```
Hashable을 채택해보셨군요! 
해당 프로토콜을 채택하신 이유가 궁금합니다 🙋‍♂️

longlivedrgn replied:
- DiffableDatasource 의 section identifier와 item identifier로 사용하기 위한 타입은 Hashable 프로토콜을 만족해야 해요. 여기서 Hashable 프로토콜을 채택하면, 해당 타입은 해쉬 값을 구할 수 있게 돼요. 해쉬 값을 갖는다는 것은 Dictionary의 Key 값이 될 수 있다는 뜻이에요. 즉, 훨씬 더 빠르게, 탐색을 할 수 있어요. 기본 Swift 타입들은 Default로 Hashable을 채택하고 있지만(+ 연관 값이 없는 enum 타입), Custom type은 그렇지 않기에 Hashable 프로토콜을 직접 채택해주었어요.

`저희는 여기서 왜 item identifier와 section identifier가 Hashable해야하는 지 유추해봤어요!`
   -  WWDC에서 나온 Diffable Datasource의 등장 이유는, reloadData(), performBatchUpdates()와 같이 통째로 데이터를 업데이트하는 것이 아닌, 변경된 데이터만 확인해서 해당 데이터만 변경하려고 나왔어요.(+성능 문제)
 - 어떤 데이터가 변경되었는 지를 확인하기 위해서는 업데이트되는 데이터 셋 속의 item들이 유일해야 해요!
        -  예) [”Miro”, “Miro”, “Jake”] ⇒ [”Miro”, “Logan”, “Jake”] 에서 무슨 “Miro”가 변경되었는지 알아야 하는데, 유일하지 않은 item이 존재하면 뭐가 변경되었는지 알 수가 없어요!
 - 그런데, Hashable을 채택하면, 거의 매번 다른 해쉬 값을 통해서 서로 다른 Identifier를 생성하므로 각각의 아이템들이 다름을 보장할 수 있어요. 또한, Hashable은 Equatable임을 보장하기에, 서로 다른 지 확인(==)할 때, 해쉬 값을 비교해서 훨씬 더 빠르게 서로 다른 지 확인할 수 있습니다.

**이를 통해서, Diffable Datasource는 Hashable한 타입으로 item identifier, section identifier를 설정하여, snapshot을 빠르게 비교할 수 있어요!**

jryoun1 replied:
> WWDC에서 나온 Diffable Datasource의 등장 이유는, reloadData(), performBatchUpdates()와 같이 통째로 데이터를 업데이트하는 것이 아닌, 변경된 데이터만 확인해서 해당 데이터만 변경하려고 나왔어요.(+성능 문제)

이 부분에 대해서는 diffableDatasource의 등장 이유는 아래의 사진과 같은 기존 collectionview, tableview에서 값이 제거되는 식의 동작이 일어날 때 index의 sync가 제대로 맞지 않아서 크래시가 발생하는 경우들이 빈번했고, 크래시는 사용자에게 좋은 경험이 아니기 때문에 이를 해결하기 위해서는 reloadData를 사용해서 sync를 맞춰줄 수 있지만, reloadData를 사용하게 되면 애니메이션이 없이 동작되기 때문에 UX적으로 사용자에게 좋은 경험을 제공하지 못하는 것 같다는 이야기가 있는 것 같아요 (애플관점).
따라서 이러한 부분을 크래시가 발생하지 않게 대신 잘 진행해주기 위해서 등장한 것으로 알고 있어서 [WWDC19: Advances in UI Data Source](https://developer.apple.com/videos/play/wwdc2019/220)를 한 번 직접 보시는 것을 추천드립니다. 
![image](https://github.com/tasty-code/ios-box-office/assets/45090197/5f40e972-a755-444e-ac6e-a6720831cf56)

넵 나머지는 부분들은 잘 확인해주신 것 같습니다 👍 😄

jryoun1 commented on BoxOffice/Model/DTO/BoxOffice.swift:
```diff
@@ -20,7 +20,7 @@ struct BoxOfficeResult: Decodable {
     let type: String
     let rangeOfDate: String
-    let dailyBoxOfficeList: [DailyBoxOfficeList]
+    let dailyBoxOfficeList: [DailyBoxOffice]
```
List를 suffix에 붙여주기 보다는 s를 사용하는 것(`dailyBoxOffices` )이 조금 더 Swift스러운 것 같습니다!

jryoun1 commented on BoxOffice/SceneDelegate.swift:
```diff
@@ -11,12 +11,16 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
-        // If using a storyboard, the `window` property will automatically be initialized and attached to the scene.
-        // This delegate does not imply the connecting scene or session are new (see `application:configurationForConnectingSceneSession` instead).
-        guard let _ = (scene as? UIWindowScene) else { return }
+        guard let windowScene = (scene as? UIWindowScene) else { return }
```
불필요한 주석제거 좋네요! 
아래 함수들도 불필요한 주석은 제거할 수 있을 것 같습니다 👍

jryoun1 commented on BoxOffice/Scenes/DailyBoxOfficeCell.swift:
```diff
@@ -0,0 +1,203 @@
+
+        static let upSymbolName: String = "arrowtriangle.up.fill"
+        static let downSymbolName: String = "arrowtriangle.down.fill"
+        static let noneChangeOfRankState = "-"
```
다른 변수들과 다르게 여기만 타입 추론이 없는 것 같습니다 🙏 
convention을 맞춰주세요!

jryoun1 commented on BoxOffice/Scenes/DailyBoxOfficeCell.swift:
```diff
@@ -0,0 +1,203 @@
+
+    static let identifier = String(describing: DailyBoxOfficeCell.self)
+
+    private enum Constants {
```
해당 셀내에서 사용하는 상수들을 전부 모아두셨네요 👍 💯

jryoun1 commented on BoxOffice/Scenes/DailyBoxOfficeCell.swift:
```diff
@@ -0,0 +1,203 @@
+        static let titleStackViewTrailingInset: CGFloat = 40.0
+    }
+
+    private let numberFormatter: NumberFormatter = {
+        let formatter = NumberFormatter()
+        formatter.numberStyle = .decimal
+
+        return formatter
+    }()
```
NumberFormatter를 사용해보셨군요 👏 
이렇게 cell내에서 선언해서 사용하는 것도 좋지만 만약 여러 cell에서 사용한다고 하면 매번 모든 cell들이 numberFomatter를 가져야할 것 같아요. 

extension을 사용해서 numberFormatter는 한 번만 만들고 원하는 형태로 반환해주도록 개선해볼까요?

longlivedrgn replied:
넵! Formatter의 extension으로 수정을 해봤는데, 제이크가 보시기에는 어떠신가요?

jryoun1 commented on BoxOffice/Scenes/DailyBoxOfficeCell.swift:
```diff
@@ -0,0 +1,203 @@
+
+        addSubview(rankVerticalStackView)
+        addSubview(titleAudienceVerticalStackView)
+    }
```
쓰는 방법은 사람마다, 팀마다 다를 것 같은데 개인적으로는 아래와 같은 순서로 쓰면 좀 더 가독성이 좋은 것 같습니다. 
```Swift
    private func configureHierarchy() {
        addSubview(rankVerticalStackView)    
        rankVerticalStackView.addArrangedSubview(rankNumberLabel)
        rankVerticalStackView.addArrangedSubview(dailyRankChangesLabel)

        addSubview(titleAudienceVerticalStackView)
        titleAudienceVerticalStackView.addArrangedSubview(movieTitleLabel)
        titleAudienceVerticalStackView.addArrangedSubview(audienceLabel)
    }
```

jryoun1 commented on BoxOffice/Scenes/DailyBoxOfficeCell.swift:
```diff
@@ -0,0 +1,203 @@
+        titleAudienceVerticalStackView.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -Constants.titleStackViewInset).isActive = true
+        titleAudienceVerticalStackView.leadingAnchor.constraint(equalTo: rankVerticalStackView.trailingAnchor, constant: .zero).isActive = true
+        titleAudienceVerticalStackView.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -Constants.titleStackViewTrailingInset).isActive = true
+    }
```
translatesAutoresizingMaskIntoConstraints는 constraints와 관련된 것일까요? 🙋‍♂️
view의 attributes와 관련된 것 같습니다! 

isActive를 매번 true하기 귀찮다면 아래와 같은 방법도 있을 것 같습니다. 
```Swift
NSLayoutConstraint.activate([
    rankVerticalStackView.widthAnchor.constraint(equalToConstant: Constants.rankStackViewWidth),
    rankVerticalStackView.topAnchor.constraint(equalTo: topAnchor, constant: Constants.rankStackViewInset),
    rankVerticalStackView.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -Constants.rankStackViewInset),
    rankVerticalStackView.leadingAnchor.constraint(equalTo: leadingAnchor, constant: Constants.rankStackViewleadingInset)
])
```

jryoun1 commented on BoxOffice/Scenes/DailyBoxOfficeCell.swift:
```diff
@@ -0,0 +1,203 @@
+        guard let audienceAccumulationNumber = Int(movie.audienceAccumulation) as? NSNumber,
+              let audienceCountNumber = Int(movie.audienceCount) as? NSNumber else { return String() }
+
+        guard let audienceAccumulation = numberFormatter.string(from: audienceAccumulationNumber ) else {
```
`numberFormatter.string(from: audienceAccumulationNumber )` 빈 공백이 있는 것 같아요 🙏

jryoun1 commented on BoxOffice/Scenes/DailyBoxOfficeCell.swift:
```diff
@@ -0,0 +1,203 @@
+              let audienceCountNumber = Int(movie.audienceCount) as? NSNumber else { return String() }
+
+        guard let audienceAccumulation = numberFormatter.string(from: audienceAccumulationNumber ) else {
+            return movie.audienceAccumulation
```
말씀하신 실패 케이스에서의 우선 String을 반환하는 코드가 여기군요.

음 여기서 만약 실패할 가능성이 있다면 String?으로 반환하고 에러인 경우에는 nil로 처리하고 실제로 text를 할당할때 nil이라면 아무것도 보여주지 않는 방법도 있을 것 같네요.
다만 서비스에서 `,를 찍어서 보여주지 못하는 것`과 `아예 안보여주는 것`과 둘 중에 어느 것이 사용자가 느꼈을 더 나을까를 고민해보면 저는 후자보다는 전자가 나을 것 같아서 두 분이 하신 것 처럼 작성했을 것 같긴하네요.

jryoun1 commented on BoxOffice/Scenes/DailyBoxOfficeCell.swift:
```diff
@@ -0,0 +1,203 @@
+    private func setOldRankChangesLabelText(with movie: DailyBoxOffice) {
+        guard let dailyRankChanges = Int(movie.dailyRankChanges) else { return }
+
+        if dailyRankChanges == 0 {
+            dailyRankChangesLabel.text = Constants.noneChangeOfRankState
+            return
+        }
+
+        if dailyRankChanges > 0 {
+            let image = UIImage(systemName: Constants.upSymbolName)?.withTintColor(.systemRed)
+            let dailytRankChangesText = generateRankChangesAttributedText(movie.dailyRankChanges,
+                                                                          with: image)
+            dailyRankChangesLabel.attributedText = dailytRankChangesText
+            return
+        }
+
+        if dailyRankChanges < 0 {
+            let image = UIImage(systemName: Constants.downSymbolName)?.withTintColor(.systemBlue)
+            let dailytRankChangesText = generateRankChangesAttributedText(movie.dailyRankChanges,
+                                                                          with: image)
+            dailyRankChangesLabel.attributedText = dailytRankChangesText
+            return
+        }
```
if, else if, else로 작성하지 않고 if 3개로 작성하신 이유가 궁금합니다 🙋‍♂️

sloDevz replied:
else 문을 쓰지 않는것이 가독성이 더 좋다는 얘기를 들었었는데 저희는 이에 동의하여 if문만을 사용해봤습니다 😊

jryoun1 replied:
그냥 궁금해서 드리는 질문입니다! 혹시 어느 부분에서 가독성이 더 좋다고 생각하셨는지가 궁금합니다 😄

jryoun1 commented on BoxOffice/Scenes/DailyBoxOfficeCell.swift:
```diff
@@ -0,0 +1,203 @@
+        }
+    }
+
+    private func generateRankChangesAttributedText(_ text: String,
+                                                   with image: UIImage?) -> NSMutableAttributedString {
+        let attachment = NSTextAttachment()
+        attachment.image = image
+
+        let rankChangesAttributedText = NSMutableAttributedString(attachment: attachment)
+        let text = NSAttributedString(string: text)
+        rankChangesAttributedText.append(text)
+
+        return rankChangesAttributedText
+    }
```
NSMutableAttributedString과 NSTextAttachment를 사용해보셨군요 👏

jryoun1 commented on BoxOffice/Scenes/DailyBoxOfficeCell.swift:
```diff
@@ -0,0 +1,203 @@
+    }
+
+
+    required init?(coder: NSCoder) {
```
이 함수는 override init(frame: CGRect) 밑에 있는 것이 비슷한 것끼리 모여있는 것 같은데 어떻게 생각하시나요?

jryoun1 commented on BoxOffice/Scenes/DailyBoxOfficeViewController.swift:
```diff
@@ -0,0 +1,156 @@
+
+    enum Constants {
+        static let formerNavigationItemTitle = "데이터 받아오는 중~!"
+    }
```
이 두 개 enum은 외부에서 접근될 일이 있을까요?

jryoun1 commented on BoxOffice/Scenes/DailyBoxOfficeViewController.swift:
```diff
@@ -0,0 +1,156 @@
+        fetchBoxOfficeData()
+    }
+
+    private func setUp() {
```
어떤 것을 setUp 해주는 것인지 함수명만 보고서는 알 수 없을 것 같아요.

jryoun1 commented on BoxOffice/Scenes/DailyBoxOfficeViewController.swift:
```diff
@@ -0,0 +1,156 @@
+    }
+
+    private func setUp() {
+        navigationItem.title = Constants.formerNavigationItemTitle
```
이 부분이 아래 기획서를 반영하신 부분일까요? 🙋‍♂️
`처음 목록을 로드할 때, 사용자에게 빈 화면만 보여주는 대신, 로드 중임을 알 수 있게 해주세요.` 

navigation title은 그대로 날짜를 작성하고 indicator만 보여줘도 되지 않을까요?

sloDevz replied:
수정하던중 indicatorView가 나오는 중에는 Title이 굳이 필요한것 같지 않아 아주 없애버렸습니다. 👍🏻

jryoun1 commented on BoxOffice/Scenes/DailyBoxOfficeViewController.swift:
```diff
@@ -0,0 +1,156 @@
+
+    private var loadingIndicatorView = UIActivityIndicatorView(style: .large)
+    private let boxOfficeManager = BoxOfficeAPIManager()
+    private var dailyBoxOfficeCollectionView: UICollectionView?
```
이 collectionView는 왜 옵셔널로 선언하셨는지 궁금합니다. 🙋‍♂️

sloDevz replied:
DailyBoxOfficeViewController 의 여러 매서드 들이 해당 dailyBoxOfficeCollectionView에 접근하기 때문에 전역으로 선언해줬고 초기화 시킬 필요가 없게 하기 위해, 그리고 ! 는 절대 사용하지 않는 원칙이 있기 때문에 옵셔널로 선언했습니다.

jryoun1 replied:
collectionView를 바로 생성해도 문제가 될 것 같은 부분이 없어서 말씀드렸습니다. (var로 선언된 이유도 비슷하게 궁금하네요)

뭔가 추후에 collectionView가 생성되어야만 하는 이유가 따로 있다면 optional이 되어야하겠지만, 지금은 그런 부분은 없는 것 같아보이고 오히려 사용할 때마다 optional binding을 해줘야하는 코드들이 불필요하게 작성되어있는 것 같습니다.

jryoun1 commented on BoxOffice/Scenes/DailyBoxOfficeViewController.swift:
```diff
@@ -0,0 +1,156 @@
+        return window
+    }
+
+    private func showIndicatorview() {
```
함수는 indicatorView를 보여줄 것 같은데 아예 window를 새롭게 만들고 거기에 view를 붙여서 보여주고 있는것 같네요 🤔
view와 window는 무엇이 다른걸까요?

jryoun1 replied:
UIView에 indicatorView붙이고 해당 view와 collectionView를 hidden처리하는 방식도 있거나, 
[bringSubviewToFront](https://developer.apple.com/documentation/uikit/uiview/1622541-bringsubviewtofront)를 사용하는 방법도 있을텐데,
window를 생성해서 view를 붙여서 보여준 이유도 궁금합니다! 🙋‍♂️

longlivedrgn replied:
window() 함수는 현재의 window를 가져오는 메소드에요! 근데, 저희가 네이밍을 애매하게 했네요🥲
그리고, 제이크의 코멘트에 대한 저희의 답변이 너무 길어질 거 같아서 아래 노션 링크에 저희가 공부한 내용을 정리해보았습니다!
감사합니다 :)
https://marred-pheasant-181.notion.site/AcitivityIndicator-View-c4e4abe61c1a4d19ac72aa0491417cd6

jryoun1 replied:
크~ 열심히 공부하셨네요 👏 💯  
다만 아직도 함수명은 show라기보다는 정확하게는 loading indicator를 view에 붙이는 함수인것 같습니다. 

물론 보여주는 방식은 지금 방식도 틀린 방식은 아니지만, show가 되려면 저 indicatorView는 항상 붙어있고 hidden이 되어있다가 특정 조건에서 isHidden이 true로 변경해주고 startAnimating을 시키는 것 정도만 포함되어야하지 않을까라는 예상이 되긴합니다. 
한 번 고민해봐주세요~ (지금 방식처럼 view를 붙였다가 remove하는 방식이 잘못된 것은 아닙니다!!)

jryoun1 commented on BoxOffice/Scenes/DailyBoxOfficeViewController.swift:
```diff
@@ -0,0 +1,156 @@
+    private func fetchBoxOfficeData() {
+        showIndicatorview()
+        let yesterDay = Date.yesterDayDateConvertToString()
+        let yesterdayDashExcepted = yesterDay.except(for: "-")
```
yesterdayWithoutDash가 좀 덜 어색해보이는 것 같습니다!

jryoun1 commented on BoxOffice/Scenes/DailyBoxOfficeViewController.swift:
```diff
@@ -0,0 +1,156 @@
+        { [weak self] data in
+            guard let boxOffice = data as? BoxOffice else { return }
+            self?.movies = boxOffice.result.dailyBoxOfficeList
+            DispatchQueue.main.async {
+                self?.navigationItem.title = yesterDay
+                self?.hideIndicatorView()
+            }
```
DispatchQueue.main.async로 감싸주신 이유가 궁금합니다~ 🙋‍♂️

sloDevz replied:
UI Update는 항상 main 스레드에서 이뤄져야 하기 때문에 UI를 Update 시키는 코드는 main 스레드에서 수행되도록 했습니다.

jryoun1 commented on BoxOffice/Scenes/DailyBoxOfficeViewController.swift:
```diff
@@ -0,0 +1,156 @@
+    }
+
+    private func configureRefreshControl() {
+        guard let dailyBoxOfficeCollectionView else { return }
+
+        dailyBoxOfficeCollectionView.refreshControl = UIRefreshControl()
+        dailyBoxOfficeCollectionView.refreshControl?.addTarget(self,
+                                                               action: #selector(handleRefreshControl),
+                                                               for: .valueChanged)
+    }
```
UIRefreshControl을 사용해보셨군요 👏
이 부분은 collectionview를 configure할 때 같이 할 수 있을 것 같아요!

jryoun1 commented on BoxOffice/Scenes/DailyBoxOfficeViewController.swift:
```diff
@@ -0,0 +1,156 @@
+    }
+
+    @objc private func handleRefreshControl() {
+        self.dailyBoxOfficeCollectionView?.refreshControl?.endRefreshing()
```
새로고침(= 서버에서 데이터를 다시 받아와서 list를 업데이트)를 해줘야할 것 같은데 지금은 그렇지 않은 것 같습니다.

longlivedrgn replied:
저희가 데이터를 다시 받아오는 걸 하지 않았네요🤔 감사합니다 :) 

그래서 최종적으로 data를 fetch하는 코드를 아래와 같이 작성하게 되었어요.
```swift
    private func fetchBoxOfficeData() {
        let yesterDay = Date.yesterDayDateConvertToString()
        let yesterdayDashExcepted = yesterDay.without("-")

        boxOfficeManager.fetchData(to: BoxOffice.self, endPoint: .boxOffice(targetDate: yesterdayDashExcepted))
        { [weak self] data in
            guard let boxOffice = data as? BoxOffice else { return }
            self?.movies = boxOffice.result.dailyBoxOffices
            DispatchQueue.main.async {
                self?.removeIndicatorView()
                self?.endRefresh()
                self?.navigationItem.title = yesterDay
            }
        }
    }
```
근데, 위 코드와 관련된 고민 점이 있어요.
- `fetchBoxOfficeData()`을 movies 프로퍼티 값만을 받아오는 역할만 시킬 것

현재 `fetchBoxOfficeData()` 함수는 API 호출 뿐만 아니라 completion handler 안에서 indicatorview를 제거하기도 하고, endRefesh를 하기도 해요. 이럴 경우, fetchBoxOfficeData()를 호출할 때마다 불필요한 함수 호출이 동반돼요.(removeIndicatorView()는 처음 실행될 때만 필요한 메소드에요.) 그래서 따로 빼서 DispatchGroup으로 묶어볼까 했는데, 이러한 저희의 고민이 의미 있는 지 궁금해요! 불필요한 함수가 계속 호출되는 게 마음에 걸리네요..

jryoun1 commented on BoxOffice/Util/Extensions/DateFormatter+.swift:
```diff
@@ -0,0 +1,27 @@
+        return formatter
+    }()
+
+    static func yesterDayDateConvertToString() -> String {
+        let date = Date()
+        guard let yesterDay = Calendar.current.date(byAdding: .day, value: -1, to: date) else { return "-" }
+        let formatter = Date.yearMonthDaySplitDashFormatter
+
+        return formatter.string(from: yesterDay)
+    }
```
잘못된 경우 -를 반환하고, -를 api endpoint의 targetDate로 넣게 되면 안 될 것 같은데 좋은 방법이 없을까요?

sloDevz replied:
의논해봤지만 마땅한 답은 찾지 못했습니다. 임의의 날짜를 넣어주는 방향도 생각해봤지만 너무 근본없는 대책인것 같아서 수정하지 않았습니다..🤔😅
특별히 제시해주실 의견이 있으신가요???

longlivedrgn replied:
저희가 error를 던져서 처리하려고 했는데, 어짜피 뒤에 fetchData에서 잘못된 targetDate가 들어올 경우, error 핸들링이 되더라고요! 그래서 에러 핸들링 말고 다른 방법이 있나 고민을 해보았는데, 마땅한 답을 찾지 못했습니다ㅜㅜ

jryoun1 commented on BoxOffice/Util/Extensions/String+.swift:
```diff
@@ -0,0 +1,19 @@
+
+import Foundation
+
+extension String {
```
extension을 잘 활용해보셨군요 💯

jryoun1 commented on BoxOffice/Scenes/DailyBoxOfficeViewController.swift:
```diff
@@ -0,0 +1,156 @@
+        snapShot.appendSections([.main])
+        snapShot.appendItems(movies)
+
+        DispatchQueue.main.async {
+            self.dataSource?.apply(snapShot)
+        }
```
이 부분도 DispatchQueue.main.async로 감싸져 있는데 이유가 궁금합니다.

sloDevz replied:
위와 동일한 이유입니다 😄 UI Update는 항상 main 스레드에서 이뤄져야 하기 때문에 UI를 Update 시키는 코드는 main 스레드에서 수행되도록 했습니다.

jryoun1 replied:
저 코드가 없었을 때 문제가 발생할까요? 🙋‍♂️

jryoun1 reviewed (changes requested):
푹쉬어야할 연휴에도 추가근무를 하셨군요 ㅎㅎ 
프로젝트 step3와 조카와의 추가근무 모두 고생하셨습니다 Team Milo 👍 
많은 부분을 공부하시고 고민하시면서 적용하신 것이 코드에 잘 녹아져 있는 것 같네요 👏💯

### 궁금한점에 대한 답변
> 1. 🔨  옵셔널 바이딩이 실패할 경우의 리턴값

해당 부분에 대해서는 코멘트에서 몇몇 케이스에 대해서 답변을 남겨두었습니다!
확인부탁드려요 🙏🏻

> 2. 🔨 dataSource.apply는 언제 호출하는게 바람직 한가?

데이터(여기서는 movies가 되겠네요)가 변경이 되었을 때마다 diffableDataSource를 통해서 collectionView가 변경이 되어야하므로 구현하신 것처럼 데이터가 변경되었을 때마다 snapShot을 apply 하는 것이 옳바른 방향이 맞습니다. 

두 분이 생각하실때는 지금 코드에서 applySnapShot()이 몇 번 호출되며, 또 어떤 경우에 호출되게 될까요? 

> 3. 🛠️ API로 부터 영화 디테일 정보를 요청하는 시점에 대하여
 
list를 그리는데 상세에 대한 response가 필요할까요? 
두 분이서 상세에 대한 데이터를 미리 받아오게 되면 장점과 단점이 있을 것 같은데 한 번 고민해보시고 답변에 남겨주세요!

jryoun1 reviewed (changes requested):
푹쉬어야할 연휴에도 추가근무를 하셨군요 ㅎㅎ 
프로젝트 step3와 조카와의 추가근무 모두 고생하셨습니다 Team Milo 👍 
많은 부분을 공부하시고 고민하시면서 적용하신 것이 코드에 잘 녹아져 있는 것 같네요 👏💯

### 궁금한점에 대한 답변
> 1. 🔨  옵셔널 바이딩이 실패할 경우의 리턴값

해당 부분에 대해서는 코멘트에서 몇몇 케이스에 대해서 답변을 남겨두었습니다!
확인부탁드려요 🙏🏻

> 2. 🔨 dataSource.apply는 언제 호출하는게 바람직 한가?

데이터(여기서는 movies가 되겠네요)가 변경이 되었을 때마다 diffableDataSource를 통해서 collectionView가 변경이 되어야하므로 구현하신 것처럼 데이터가 변경되었을 때마다 snapShot을 apply 하는 것이 옳바른 방향이 맞습니다. 

두 분이 생각하실때는 지금 코드에서 applySnapShot()이 몇 번 호출되며, 또 어떤 경우에 호출되게 될까요? 

> 3. 🛠️ API로 부터 영화 디테일 정보를 요청하는 시점에 대하여
 
list를 그리는데 상세에 대한 response가 필요할까요? 
두 분이서 상세에 대한 데이터를 미리 받아오게 되면 장점과 단점이 있을 것 같은데 한 번 고민해보시고 답변에 남겨주세요!

sloDevz committed:
refactor: 리뷰반영
- dailyBoxOfficeList -> dailyBoxOffices 로 변경
- SceneDelegate와 AppDelegate의 불필요한 주석 제거
- noneChangeOfRankState 타입 추론 가능하도록 타입명시

longlivedrgn committed:
chore: 리뷰 코멘트 반영

- translatesAutoresizingMaskIntoConstraints 코드 위치 변경
- Formatter extension을 활용하기

longlivedrgn committed:
Merge branch 'main' of https://github.com/longlivedrgn/ios-box-office

sloDevz committed:
refactor: 리뷰반영
- generateAudienceLabelText(with:) 메서드 개선
- required init?(coder: NSCoder) 위치변경
- enum private
- setUp -> setUpUI 메서드명 변경

longlivedrgn committed:
chore: String extension 변경하기

sloDevz committed:
refactor: 리뷰반영
- refreshControl 선언 위치 변경 및 호출위치 개선
- indicatorView 호출 위치 변경

longlivedrgn committed:
chore: navigation title 제거 및 DateFormatter 수정

sloDevz committed:
refactor: 리뷰반영
- indicatorView를 window가 아닌 View위에 구현
- refreshControl 의 endRefreshing()을 메서드로 묶음

longlivedrgn commented:
### 🔨 dataSource.apply는 언제 호출하는게 바람직 한가?

*데이터(여기서는 movies가 되겠네요)가 변경이 되었을 때마다 diffableDataSource를 통해서 collectionView가 변경이 되어야하므로 구현하신 것처럼 데이터가 변경되었을 때마다 snapShot을 apply 하는 것이 옳바른 방향이 맞습니다.*

*두 분이 생각하실때는 지금 코드에서 applySnapShot()이 몇 번 호출되며, 또 어떤 경우에 호출되게 될까요?*

- 맨 처음 load가 될 때, snapshot을 apply하고 RefreshControl을 통해서 다시 data를 받아올 때 apply SnapShot을 호출합니다.

저희가 생각했을 때 지금 코드에서는 위와 같은 경우 밖에 없을 것이라고 생각이 들었습니다! 현재의 코드와 같이 apply를 하는 것이 옳바른 방향인 것 같았습니다!

### 🔨 추가적인 답변

- 그 외에 저희의 궁금점 및 답변들은 위의 코멘트의 댓글로 달아두었습니다!

sloDevz commented:
### 🛠️ API로 부터 영화 디테일 정보를 요청하는 시점에 대하여

*list를 그리는데 상세에 대한 response가 필요할까요?
두 분이서 상세에 대한 데이터를 미리 받아오게 되면 장점과 단점이 있을 것 같은데 한 번 고민해보시고 답변에 남겨주세요!*

- 하루의 BoxOffice에 대한 데이터 전체를 미리 받아오게 된다면 앱을 켤때마다 BoxOffice 리스트의 정보와 각 영화들의 디테일에 대한 데이터 요청을 앱이 시작될때 한번에 하기 때문에 로딩시간이 길어지는 단점이 있다고 생각했습니다. 반대로 장점은 그 이후로부터는 더이상 지연시간없이 빠른 사용자경험을 줄 수 있다고 생각했습니다.
- 장단점을 비교하며 앱 API호출 방식을 설계할때 어떤방향이 합리적인가에 대한 결정에 가장 영향을 주는 요소는 해당 앱이 존재하는 근본적인 목적에있다는 생각을 했습니다.
- 유저가 해당 앱에 오랫동안 머물며 View 전환도 잦을 경우 앱이 실행될때 한번에 앞으로 필요할지도 모르는 데이터를 받아오는 선택지가 더 합리적일 수 있지만 그렇지 않을 경우엔 필요한 데이터를 그때그때 받아오는게 마땅하다고 판단했습니다.

jryoun1 commented on BoxOffice/AppDelegate.swift:
```diff
@@ -10,27 +10,18 @@ import UIKit
-
-
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
-        // Override point for customization after application launch.
```
👏

jryoun1 commented on BoxOffice/Util/Extensions/Formatter+.swift:
```diff
@@ -0,0 +1,26 @@
+
+import Foundation
+
+extension Formatter {
```
Formatter도 따로 분리하셨군요 👏

jryoun1 commented on BoxOffice/Util/Extensions/Int+.swift:
```diff
@@ -0,0 +1,18 @@
+
+extension Int {
+
+    var decimalizedString: String {
```
💯

jryoun1 commented on BoxOffice/Util/Extensions/DateFormatter+.swift:
```diff
@@ -0,0 +1,20 @@
+
+    static func yesterDayDateConvertToString() -> String {
+        let date = Date()
+        guard let yesterDay = Calendar.current.date(byAdding: .day, value: -1, to: date) else { return "20220324" }
```
apimanager에서 targetDate를 넘겼을 때 혹시 에러처리가 되는 코드가 어디 부분일까요? 

그와 별개로 이 부분은 사실 정답은 없는 것 같고 처리하는 방식에 따라서 다를것 같긴한데, 
- api를 호출하기 전에 막을지 
- api를 호출하고나서 다른 값이 왔을 때 보여줄지
에 따라서 결정하시면 될 것 같습니다. 

아예 targetDate 자체가 틀렸다는 것을 클라이언트에서 판단할 수 있는데 따로 판단하지 않고 서버에게 책임을 넘기는 것보다는 가능하다면 아예 api를 호출하지 않는 전자 방식이 나을 것 같습니다.
이렇게 처리하는 방법도 여러가지가 있겠지만 String?을 반환하면서 nil을 반환하고 nil인 경우에는 alertController를 띄워서 에러가 있다는 것을 알린다던지 하는 방식이 있긴하겠네요!

jryoun1 reviewed (approved):
고생하셨습니다!
제가 드린 질문들에 답변과 공부한 내용들을 잘 담아주셨던 것 같네요 👍 

몇가지 코멘트를 남기긴했지만 해당 부분들을 추후 스텝을 진행하면서 고민해보고 적용해주셔도 될 것 같아서 여기서 머지하도록 하겠습니다! 

## 질문들 모음

> 1. 🔨 dataSource.apply는 언제 호출하는게 바람직 한가?
데이터(여기서는 movies가 되겠네요)가 변경이 되었을 때마다 diffableDataSource를 통해서 collectionView가 변경이 되어야하므로 구현하신 것처럼 데이터가 변경되었을 때마다 snapShot을 apply 하는 것이 옳바른 방향이 맞습니다.

두 분이 생각하실때는 지금 코드에서 applySnapShot()이 몇 번 호출되며, 또 어떤 경우에 호출되게 될까요?

맨 처음 load가 될 때, snapshot을 apply하고 RefreshControl을 통해서 다시 data를 받아올 때 apply SnapShot을 호출합니다.
저희가 생각했을 때 지금 코드에서는 위와 같은 경우 밖에 없을 것이라고 생각이 들었습니다! 현재의 코드와 같이 apply를 하는 것이 옳바른 방향인 것 같았습니다!

👏

> 2. 🛠️ API로 부터 영화 디테일 정보를 요청하는 시점에 대하여
list를 그리는데 상세에 대한 response가 필요할까요?
두 분이서 상세에 대한 데이터를 미리 받아오게 되면 장점과 단점이 있을 것 같은데 한 번 고민해보시고 답변에 남겨주세요!

하루의 BoxOffice에 대한 데이터 전체를 미리 받아오게 된다면 앱을 켤때마다 BoxOffice 리스트의 정보와 각 영화들의 디테일에 대한 데이터 요청을 앱이 시작될때 한번에 하기 때문에 로딩시간이 길어지는 단점이 있다고 생각했습니다. 반대로 장점은 그 이후로부터는 더이상 지연시간없이 빠른 사용자경험을 줄 수 있다고 생각했습니다.
장단점을 비교하며 앱 API호출 방식을 설계할때 어떤방향이 합리적인가에 대한 결정에 가장 영향을 주는 요소는 해당 앱이 존재하는 근본적인 목적에있다는 생각을 했습니다.
유저가 해당 앱에 오랫동안 머물며 View 전환도 잦을 경우 앱이 실행될때 한번에 앞으로 필요할지도 모르는 데이터를 받아오는 선택지가 더 합리적일 수 있지만 그렇지 않을 경우엔 필요한 데이터를 그때그때 받아오는게 마땅하다고 판단했습니다.

이 부분에 대해서는 어떤 의도에서 그렇게 생각하셨는지는 이해가 되었습니다. 
다만 사용자가 들어갈지 안들어갈지도 모르는 상세의 데이터까지 모두 받아오느냐고 list를 보여주는것이 늦어지는 것이 오히려 사용자를 불편하게 만드는 부분이 아닐까라고 생각이 듭니다.
또한 데이터를 미리 받게된다면 실시간성으로 반영이 되지 않을 수 있기 때문에 sync가 다른 부분을 보여줄 수도 있을 것 같고요. 

그래서 말씀하신 것처럼 물론 앱의 목적에 따라서 다르긴 하겠지만 대부분의 앱에서는 실제 필요할 때 해당 api를 호출하도록 하는 것 같습니다. 만약 너무 자주 호출해야하거나, 받아오는 데이터의 양이 많거나 하는 경우에는 캐싱이라는 부분도 있으니 추후에 한 번 공부해보시면 좋을 것 같아요!
